active_discussion/representation.md

```diff
@@ -1,14 +1,152 @@
+the validity invariant to hold **at all times** and is thus allowed to use
+these invariants to (e.g.) affect the layout of data structures or do other
+optimizations.
+
+Therefore, the validity invariant must **at minimum** justify all the
+layout optimizations that the compiler does. We may want a stronger
+invariant, however, so as to leave room for future optimization.
+
+As an example, a value of `&T` type can never be null -- therefore,
+`Option<&T>` can use null to represent `None`.
+
+## Goals
+
+- Define what we guarantee about the layout of various types
+  and the effect of `#[repr]` annotations.
+- Define the **validity requirements** of various types. These are the
+  requirements that must hold at all times when the compiler considers
+  a value to be initialized.
+    - Also examine when/how we could dynamically check these requirements.
+- Uncover the sorts of constraints that we may wish to satisfy in the
+  future.
+
+## Some interesting examples and questions
+
+- `&T` where `T: Sized`
+  - This is **guaranteed** to be a non-null pointer
+- `Option<&T>` where `T: Sized`
+  - This is **guaranteed** to be a nullable pointer
+- `Option<extern "C" fn()>`
+- `usize`
+  - Platform dependent size, but guaranteed to be able to store a pointer?
+  - Also an array length?
+- Uninitialized bits -- for which types are uninitialized bits valid?
+- If you have `struct A { .. }` and `struct B { .. }` with no
+  `#[repr]` annotations, and they have the same field types, can we
+  say that they will have the same layout?
+  - or do we have the freedom to rearrange the types of `A` but not
+    `B`, e.g. based on PGO results
+- Rust currently says that no single value may be larger than `isize` bytes
+  - is this good? can it be changed? does it matter *here* anyway?
+
+## Active threads
+
+To start, we will create threads for each major categories of types
+(with a few suggested focus points):
+
+- Integers and floating points
+    - What about uninitialized values?
+- Booleans
+    - Prior discussions ([#46156][], [#46176][]) documented bool as a single
+      byte that is either 0 or 1.
+- Enums
+    - See dedicated thread about "niches" and `Option`-style layout optimization
+      below.
+    - Define: C-like enum
+    - Can a C-like enum ever have an invalid discriminant? (Presumably not)
+    - Empty enums and the `!` type
+    - [RFC 2195][] defined the layout of `#[repr(C)]` enums with payloads.
+    - [RFC 2363][] offers a proposal to permit specifying discriminations.
+- Structs
+    - Do we ever say *anything* about how a `#[repr(rust)]` struct is laid out
+      (and/or treated by the ABI)?
+      - e.g., what about different structs with same definition
+      - across executions of the same program?
+- Tuples
+    - Are these effectively anonymous structs? 
+- Unions
+    - Can we ever say anything about the initialized contents of a union?
+    - Is `#[repr(C)]` meaningful on a union?
```

Comment:
It should guarantee that all fields have the same address. This might be the case without `#[repr(C)]`.

Reply:
added